extensions/native/guest-macro/src/transportation.rs

```diff
@@ -0,0 +1,321 @@
+}
+
+impl<F: Field> MacroInstruction<F> {
+    fn literal(instruction: Instruction<F>) -> Self {
+        Self {
+            opcode: instruction.opcode,
+            operands: instruction.operands().into_iter().map(Literal).collect(),
+        }
+    }
+
+    fn new<const N: usize>(opcode: VmOpcode, operands: [Operand<F>; N]) -> Self {
+        Self {
+            opcode,
+            operands: operands.to_vec(),
+        }
+    }
+}
+
+/*
+
+how things are going to go:
+
+transportation will provide some MacroInstructions
+
+the body will be converted to MacroInstructions
+
+MacroInstructions are converted to an asm! call
+ */
+
+pub fn compiled_kernel_to_function<F: PrimeField32>(
+    compiled_kernel: CompiledKernel<F>,
+) -> TokenStream {
+    let mut instructions = vec![];
+    let mut input_vars = vec![];
+    let return_name = "result".to_string();
+
+    let function_name_token = format_ident!("{}", compiled_kernel.function_name);
+    let return_type_token = format_ident!("{}", compiled_kernel.rust_return_type);
+    let return_name_token = format_ident!("{}", return_name);
+
+    let mut arguments = vec![];
+
+    for argument in compiled_kernel.arguments {
+        let var_name = "var_".to_string() + &argument.name;
+        let var_name_token = format_ident!("{}", var_name);
+        let rust_type_token = format_ident!("{}", argument.rust_type);
+        arguments.push(quote! {
+            #var_name_token: #rust_type_token
+        });
+
+        input_vars.push(var_name);
+        instructions.extend(transport_rust_to_edsl(
+            argument.rust_type,
+            argument.edsl_type,
+            "var_".to_string() + &argument.name,
+            argument.fp,
+        ));
+    }
+
+    instructions.extend(
+        compiled_kernel
+            .body
+            .into_iter()
+            .map(MacroInstruction::literal),
+    );
+    instructions.extend(transport_edsl_to_rust(
+        compiled_kernel.rust_return_type,
+        compiled_kernel.edsl_return_type,
+        return_name.clone(),
+        compiled_kernel.return_fp,
+    ));
+
+    let mut instructions_string = String::new();
+    for instruction in instructions.iter() {
+        instructions_string.push_str(&format!("{:?}\n", instruction));
+    }
+    std::fs::write("instructions.txt", instructions_string).expect("Failed to write file");
+
+    let asm_call: TokenStream =
+        instructions_to_asm_call(instructions, input_vars, vec![return_name.clone()])
+            .parse()
+            .unwrap();
+
+    quote! {
+        fn #function_name_token(#(#arguments),*) -> #return_type_token {
+            let #return_name_token: #return_type_token;
+            #asm_call
+            #return_name_token
+        }
+    }
+}
+
+fn u32_to_directive(x: u32) -> String {
+    let opcode = x & 0b1111111;
+    let funct3 = (x >> 12) & 0b111;
+    let rd = (x >> 7) & 0b11111;
+    let rs1 = (x >> 15) & 0b11111;
+    let mut simm12 = (x >> 20) as i32;
```

Comment:
the .insn takes a signed number?

Reply:
most insn string will be meaningless actually, why not directly use ".4byte 0x00000293" like raw string?

Reply:
according to https://sourceware.org/binutils/docs/as/RISC_002dV_002dDirectives.html
> But in fact, the instruction formats are difficult to use for some users, so most of them are using ‘.word’ to encode the instruction directly, rather than using ‘.insn’. It is fine for now, but will be wrong when the mapping symbols are supported, since ‘.word’ will not be shown as an instruction, it should be shown as data. Therefore, we also support two more formats of the ‘.insn’, the instruction ‘add a0, a1, a2’ could also be written as ‘.insn 0x4, 0xc58533’ or ‘.insn 0xc58533’. When the insn_length is set, then assembler will check if the value is a valid insn_length bytes instruction.

@TlatoaniHJ we can try, I'm not sure the rust `asm!` macro supports these variations of the directive

Reply:
I checked the RISC-V ELF spec somewhere, and it seems mapping symbols are added, but I'm not sure it is really used yet

Reply:
it seems rust does not support these nice .insn directives